cmake/CompilerWarnings.cmake

```diff
@@ -0,0 +1,88 @@
+        /w14905 # wide string literal cast to 'LPSTR'
+        /w14906 # string literal cast to 'LPWSTR'
+        /w14928 # illegal copy-initialization; more than one user-defined conversion has been implicitly applied
+        /permissive- # standards conformance mode for MSVC compiler.
+
+        # Disables, remove when appropriate
+        /wd4996 # disable warnings about deprecated functions
+    )
+
+    set(CLANG_WARNINGS
+        -Wall
+        -Wextra # reasonable and standard
+        -Wshadow # warn the user if a variable declaration shadows one from a parent context
+        -Wnon-virtual-dtor # warn the user if a class with virtual functions has a non-virtual destructor. This helps catch hard to track down memory errors
+        -Wold-style-cast # warn for c-style casts
+        -Wcast-align # warn for potential performance problem casts
+        -Wunused # warn on anything being unused
+        -Woverloaded-virtual # warn if you overload (not override) a virtual function
+        -Wpedantic # warn if non-standard C++ is used
+        -Wconversion # warn on type conversions that may lose data
+        -Wsign-conversion # warn on sign conversions
+        -Wnull-dereference # warn if a null dereference is detected
+        -Wdouble-promotion # warn if float is implicit promoted to double
+        -Wformat=2 # warn on security issues around functions that format output (ie printf)
+    )
+
+    if(WARNINGS_AS_ERRORS)
+        set(CLANG_WARNINGS ${CLANG_WARNINGS} -Werror)
+        set(MSVC_WARNINGS ${MSVC_WARNINGS} /WX)
+    endif()
+
+    set(GCC_WARNINGS
+        ${CLANG_WARNINGS}
+        -Wmisleading-indentation # warn if indentation implies blocks where blocks do not exist
+        -Wduplicated-cond # warn if if / else chain has duplicated conditions
+        -Wlogical-op # warn about logical operations being used where bitwise were probably wanted
+        -Wuseless-cast # warn if you perform a cast to the same type
+    )
+
+    # Don't enable -Wduplicated-branches for GCC < 8.1 since it will lead to false positives
+    # https://github.com/gcc-mirror/gcc/commit/6bebae75035889a4844eb4d32a695bebf412bcd7
+    if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 8.1)
+        set(GCC_WARNINGS
+            ${GCC_WARNINGS}
+            -Wduplicated-branches # warn if if / else branches have duplicated code
+        )
+    endif()
+
+    if(MSVC)
+        set(FILE_WARNINGS ${MSVC_WARNINGS})
+    elseif(CMAKE_CXX_COMPILER_ID MATCHES ".*Clang")
+        set(FILE_WARNINGS ${CLANG_WARNINGS})
+    elseif(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+        set(FILE_WARNINGS ${GCC_WARNINGS})
+    else()
+        message(AUTHOR_WARNING "No compiler warnings set for '${CMAKE_CXX_COMPILER_ID}' compiler.")
+    endif()
+
+    foreach(WARNING ${FILE_WARNINGS})
+        set_property(SOURCE ${ARGV} APPEND_STRING PROPERTY COMPILE_FLAGS " ${WARNING}")
```

Comment:
Shouldn't this be `target_compile_options(${TARGET} PRIVATE ${WARNING})` or maybe even a way to turn the list into space separated string.

Reply:
As the name of the function implies, we want to set warnings at a file-level. This is the least complicated way of being able to control which files we care about and which we don't.

Reply:
Hmm not sure if this works as expected, as the Windows header is still generating errors:

> `C:\Program Files (x86)\Windows Kits\8.1\include\um\combaseapi.h(229): error C2760: syntax error: unexpected token 'identifier', expected 'type specifier'`

https://ci.sfml-dev.org/#/builders/26/builds/94/steps/6/logs/stdio

Reply:
Yes... `set_property(SOURCE ... PROPERTY COMPILE_FLAGS ...)` can only set compiler flags that are used per compiler invocation, which is per TU (.cpp file after pre-processing). In order to exclude headers or disable warnings for specific headers (at least in MSVC) there is no other way than to wrap the corresponding header `#include`s in `#pragma`s. Rather painful process, but I've done it for my own projects often enough that I know you can survive it. 😁

Reply:
Do you think, we can make use of `/external` for the VS versions that support it? 

https://devblogs.microsoft.com/cppblog/broken-warnings-theory/

Reply:
Not a fan of relying on stuff that MS still considers as experimental. Until it is no longer experimental, it might break in random ways without near-term commitment from MS for a fix, see: https://gitlab.kitware.com/cmake/cmake/-/issues/17904

Reply:
It looks like this isn't even because of any or the warnings, but because of `/permissive`.
Can you and do we want to update the Windows SDK on the Windows build server?

See also: https://github.com/microsoft/krabsetw/issues/42

Reply:
I think we should just remove `/permissive-` for now. I don't want to think of the headaches supporting/guiding people that have issues because they are still using "broken" SDK versions.